### PR TITLE
COS-23: Invoice Handling

### PR DESCRIPTION
**Pull request by awg-github**
1. If in Odoo, there is no existing invoice whose account_invoice.x_civicrm_id matches with contribution_id, a new invoice is created according to the mapping for "Contributions" entity on contribution tab in mapping sheet.
cos-23-1

1. If in Odoo, there is one or more existing invoice whose account_invoice.x_civicrm_id matches with contribution_id, check the last invoice: 	
   1. if the invoice is not posted, do "Line Items Handling".
   1. if the invoice is posted, check the if exact same invoices lines exist in the last matched invoice in Odoo as per CiviCRM contribution: 		
      1. if yes, do "Status and Payment Handling"
      1. if no, unreconcile and cancel the invoice. Create a new one and do "Line Items Handling". Re-reconcile the payments with the new invoice after it is posted and before Status and Payment Handling.
cos-23-2

### CODE REVIEW

**Comment by guanhuan** [comment]
Hi @awg-github , the gif looks good. However, I think it only covers point 1. Are the specs in point 2 included in this PR too or is there another PR coming for this ticket?

File: data/product_data.xml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        <record model="product.product" id="data_product_CVBK">
+            <field name="name">CiviCRM Booking</field>
+            <field name="default_code">CVBK</field>
+        </record>
+        <record model="product.product" id="data_product_CVCTB">
+            <field name="name">CiviCRM Contribution</field>
+            <field name="default_code">CVCTB</field>
+        </record>
+    </data>
+</odoo>
```
Comment (omarabuhussein) [inline data/product_data.xml]: please add space at the end of  the file

File: models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+    'contact_civicrm_id': ('res.partner', 'x_civicrm_id', 'partner_id'),
+    'account_code': ('account.account', 'code', 'account_id'),
+    'currency_code': ('res.currency', 'name', 'currency_id'),
+    'product_code': ('product.product', 'default_code', 'product_id'),
+    'tax_name': ('account.tax', 'name', 'invoice_line_tax_ids'),
+    'journal_code': ('account.journal', 'code', 'journal_id'),
+    'invoice_civicrm_id': ('account.invoice', 'x_civicrm_id', 'x_civicrm_id'),
+    'invoice_line_civicrm_id': ('account.invoice.line', 'x_civicrm_id',
+                                'x_civicrm_id'),
+    # 'payment_civicrm_id': ('account.payment', 'x_civicrm_id', 'x_civicrm_id'),
```
Comment (omarabuhussein) [inline models/account_invoice.py]: please remove the commented out code.

File: models/account_invoice.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+class AccountInvoice(models.Model):
+    _inherit = "account.invoice"
+
+    x_civicrm_id = fields.Integer(string='Civicrm Id', required=False,
+                                  help='Civicrm Id')
+
+    @api.model
+    def civicrm_sync(self, input_params):
+        """Synchronizes CiviCRM Contributions to Odoo invoice.
+         Creates new invoice if not exists and updates is it is
```
Comment (omarabuhussein) [inline models/account_invoice.py]: typo, it should be : 

and updates it if it is not present

File: models/account_invoice.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+         present in Odoo. Returns back to CiviCRM assigned invoice_id and
+         update_date and data processing status.
+        """
+        _logger.debug('Start Invoice civicrm_sync')
```
Comment (omarabuhussein) [inline models/account_invoice.py]: I don't think this is a clear debugging info, maybe **Start CiviCRM contribution to invoice syncing** is better message ?

File: models/account_invoice.py [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+        invoice = self.with_context(active_test=False).search(
+            [('x_civicrm_id', '=', x_civicrm_invice_id)], order='id desc',
+            limit=1)
+        _logger.debug('invoice = {}'.format(invoice))
+        invoice_state = invoice.state
+        _logger.debug('invoice_state = {}'.format(invoice_state))
+
+        try:
+
+            # Create and post new invoise if not exist
```
Comment (omarabuhussein) [inline models/account_invoice.py]: typo in word : invoise

**Comment by awg-github** [comment]
@omarabuhussein, please, check our fixes

**Pull request merged**